File: mobile/lib/prompts/v5_artist_weight_repair.dart
import 'dart:math';

import '../artist/artist_recipe.dart';
import '../artist/curated_artists.dart';

class V5ArtistWeightRepairResult {
  final String output;
  final int adjustedWeightedGroups;
  final int wrappedUnweightedArtists;
  final int totalAdjusted;
  final int artistTagCount;
  final int qualityTagCount;
  final int otherTagCount;

  const V5ArtistWeightRepairResult({
    required this.output,
    required this.adjustedWeightedGroups,
    required this.wrappedUnweightedArtists,
    required this.totalAdjusted,
    required this.artistTagCount,
    required this.qualityTagCount,
    required this.otherTagCount,
  });
}

/// Community migration reports for early NAI Diffusion V5 commonly suggest
/// starting at roughly one third to one half of a V4.5 prompt weight. NovelAI
/// does not publish an official conversion ratio.
const double minV5ArtistRepairMultiplier = 1 / 3;
const double maxV5ArtistRepairMultiplier = 1 / 2;
const double defaultV5ArtistDrawMin = .2;
const double defaultV5ArtistDrawMax = 1.2;

class _PromptTag {
  final String raw;
  final String value;
  final double weight;
  final String kind;
  final bool sourceWeighted;

  const _PromptTag({
    required this.raw,
    required this.value,
    required this.weight,
    required this.kind,
    required this.sourceWeighted,
  });

  _PromptTag copyWithWeight(double nextWeight) => _PromptTag(
        raw: raw,
        value: value,
        weight: nextWeight,
        kind: kind,
        sourceWeighted: sourceWeighted,
      );
}

final RegExp _qualityPattern = RegExp(
  r'^(?:masterpiece|best[_ ]quality|amazing[_ ]quality|very[_ ]aesthetic|extremely[_ ]detailed(?:[_ ]cg)?|ultra[-_ ]?detailed|high[_ ]quality|great[_ ]quality|good[_ ]quality|average[_ ]quality|low[_ ]quality|worst[_ ]quality|aesthetic|very[_ ]pleasing|no[_ ]text)$',
  caseSensitive: false,
);
final RegExp _yearPattern = RegExp(r'^year[_ ]?\d{4}$', caseSensitive: false);
final RegExp _negativePattern = RegExp(
  r'^(?:no\s+|negative\s+|avoid\s+)',
  caseSensitive: false,
);
final RegExp _stylePattern = RegExp(
  r'(?:style|realism|impasto|illustration|painting|lineart|lighting|shading|\bcg\b|photorealistic|monochrome|sketch|watercolor|brushstroke|anime coloring|game cg)',
  caseSensitive: false,
);
final RegExp _numericScopePattern = RegExp(
  r'([+-]?(?:\d+(?:\.\d*)?|\.\d+))\s*::([\s\S]*?)::',
);

final Set<String> _knownBareArtistTags = {
  ...kCuratedArtistTags.map((artist) => artist.name),
  'xiaoluo_xl',
};

String _formatWeight(double value) {
  final rounded = (value * 100).round() / 100;
  var fixed = (rounded == -0.0 ? 0.0 : rounded).toStringAsFixed(2);
  while (fixed.contains('.') && fixed.endsWith('0')) {
    fixed = fixed.substring(0, fixed.length - 1);
  }
  if (fixed.endsWith('.')) fixed = fixed.substring(0, fixed.length - 1);
  return fixed;
}

String _normalizeCompatibilityAscii(String input) {
  final buffer = StringBuffer();
  for (final rune in input.runes) {
    if (rune >= 0xff01 && rune <= 0xff5e) {
      buffer.writeCharCode(rune - 0xfee0);
    } else if (rune == 0x3000) {
      buffer.write(' ');
    } else {
      buffer.writeCharCode(rune);
    }
  }
  return buffer.toString();
}

String _normalizePromptSource(String input) =>
    _normalizeCompatibilityAscii(input)
        .replaceAll('\u00a0', ' ')
        .replaceAllMapped(
          RegExp(r'\\([_()[\]{}])'),
          (match) => match.group(1)!,
        )
        .replaceAll(RegExp(r'[，、；;\r\n]+'), ',');

String _cleanPromptValue(String raw) {
  var value = raw.trim();
  while (value.startsWith('::')) {
    value = value.substring(2).trimLeft();
  }
  while (value.endsWith('::')) {
    value = value.substring(0, value.length - 2).trimRight();
  }
  return value.trim();
}

String? _repairKnownBareArtistName(String raw) {
  final candidate = canonicalArtistTagName(raw);
  if (_knownBareArtistTags.contains(candidate)) return candidate;
  final opens = '('.allMatches(candidate).length;
  final closes = ')'.allMatches(candidate).length;
  final repaired = '$candidate)';
  if (opens == closes + 1 && _knownBareArtistTags.contains(repaired)) {
    return repaired;
  }
  return null;
}

({String value, String kind})? _classifyAndNormalizeValue(String raw) {
  final value = _cleanPromptValue(raw);
  if (value.isEmpty) return null;

  final explicitArtist = RegExp(
    r'^artist\s*:\s*([\s\S]+)$',
    caseSensitive: false,
  ).firstMatch(value);
  if (explicitArtist != null) {
    final rawName = explicitArtist.group(1)!.trim();
    final name =
        _repairKnownBareArtistName(rawName) ?? canonicalArtistTagName(rawName);
    return name.isEmpty ? null : (value: 'artist:$name', kind: 'artist');
  }

  final knownArtist = _repairKnownBareArtistName(value);
  if (knownArtist != null) {
    return (value: 'artist:$knownArtist', kind: 'artist');
  }

  final comparable = value.replaceAll('_', ' ').trim();
  if (_yearPattern.hasMatch(value)) return (value: value, kind: 'year');
  if (_qualityPattern.hasMatch(value)) return (value: value, kind: 'quality');
  if (_negativePattern.hasMatch(comparable)) {
    return (value: value, kind: 'negative');
  }
  if (_stylePattern.hasMatch(comparable)) return (value: value, kind: 'style');
  return (value: value, kind: 'other');
}

_PromptTag? _parseTerm(
  String raw,
  double? inheritedWeight,
  int legacyLevel,
) {
  var value = _cleanPromptValue(raw);
  if (value.isEmpty) return null;
  var baseWeight = inheritedWeight ?? 1.0;
  var sourceWeighted = inheritedWeight != null || legacyLevel != 0;

  final parenthesizedWeight = RegExp(
    r'^\(\s*([\s\S]+)\s*:\s*([+-]?(?:\d+(?:\.\d*)?|\.\d+))\s*\)$',
  ).firstMatch(value);
  if (parenthesizedWeight != null) {
    value = parenthesizedWeight.group(1)!.trim();
    baseWeight = double.parse(parenthesizedWeight.group(2)!);
    sourceWeighted = true;
  } else if (value.startsWith('(') && value.endsWith(')')) {
    value = value.substring(1, value.length - 1).trim();
  }

  final normalized = _classifyAndNormalizeValue(value);
  if (normalized == null) return null;
  final weight = baseWeight * pow(1.05, legacyLevel).toDouble();
  return _PromptTag(
    raw: raw,
    value: normalized.value,
    weight: weight.isFinite ? weight : 1,
    kind: normalized.kind,
    sourceWeighted: sourceWeighted,
  );
}

List<_PromptTag> _parseTermList(String source, double? inheritedWeight) {
  final output = <_PromptTag>[];
  final token = StringBuffer();
  var curly = 0;
  var square = 0;
  int? startingLevel;

  void flush() {
    final parsed = _parseTerm(
      token.toString(),
      inheritedWeight,
      startingLevel ?? 0,
    );
    if (parsed != null) output.add(parsed);
    token.clear();
    startingLevel = null;
  }

  for (final rune in source.runes) {
    final char = String.fromCharCode(rune);
    if (char == '{') {
      curly++;
      continue;
    }
    if (char == '}') {
      curly = max(0, curly - 1);
      continue;
    }
    if (char == '[') {
      square++;
      continue;
    }
    if (char == ']') {
      square = max(0, square - 1);
      continue;
    }
    if (char == ',') {
      flush();
      continue;
    }
    if (startingLevel == null && char.trim().isNotEmpty) {
      startingLevel = curly - square;
    }
    token.write(char);
  }
  flush();
  return output;
}

List<_PromptTag> _parsePromptTags(String input) {
  final source = _normalizePromptSource(input);
  final output = <_PromptTag>[];
  var cursor = 0;
  for (final match in _numericScopePattern.allMatches(source)) {
    output.addAll(_parseTermList(source.substring(cursor, match.start), null));
    final weight = double.tryParse(match.group(1)!) ?? 1;
    output.addAll(_parseTermList(match.group(2)!, weight));
    cursor = match.end;
  }
  output.addAll(_parseTermList(source.substring(cursor), null));
  return output;
}

String _formatPromptTags(Iterable<_PromptTag> tags) => tags
    .map((tag) => '${_formatWeight(tag.weight)}::${tag.value} ::')
    .join(', ');

V5ArtistWeightRepairResult _createRepairResult(List<_PromptTag> tags) {
  final artistTagCount = tags.where((tag) => tag.kind == 'artist').length;
  final qualityTagCount = tags.where((tag) => tag.kind == 'quality').length;
  return V5ArtistWeightRepairResult(
    output: _formatPromptTags(tags),
    adjustedWeightedGroups: tags.where((tag) => tag.sourceWeighted).length,
    wrappedUnweightedArtists:
        tags.where((tag) => tag.kind == 'artist' && !tag.sourceWeighted).length,
    totalAdjusted: tags.length,
    artistTagCount: artistTagCount,
    qualityTagCount: qualityTagCount,
    otherTagCount: tags.length - artistTagCount - qualityTagCount,
  );
}

double _normalizedRandomSample(double Function() random) {
  final value = random();
  return value.isFinite ? value.clamp(0, 1).toDouble() : .5;
}

double _randomRepairMultiplier(double Function() random) =>
    minV5ArtistRepairMultiplier +
    _normalizedRandomSample(random) *
        (maxV5ArtistRepairMultiplier - minV5ArtistRepairMultiplier);

List<_PromptTag> _migratePromptTags(
  List<_PromptTag> tags,
  double Function() random, {
  ({double min, double max})? bounds,
}) =>
    tags.map((tag) {
      var migrated = tag.weight * _randomRepairMultiplier(random);
      if (bounds != null) migrated = migrated.clamp(bounds.min, bounds.max);
      return tag.copyWithWeight((migrated * 100).round() / 100);
    }).toList();

ArtistRecipe _recipeFromTags(
  List<_PromptTag> tags,
  int index,
  String idPrefix,
) {
  final prompt = _formatPromptTags(tags);
  final artists = tags
      .where((tag) => tag.kind == 'artist')
      .map((tag) => canonicalArtistTagName(
            tag.value.replaceFirst(
              RegExp(r'^artist\s*:', caseSensitive: false),
              '',
            ),
          ))
      .where((name) => name.isNotEmpty)
      .toList();
  final id = '$idPrefix-${index + 1}-${DateTime.now().microsecondsSinceEpoch}';
  return ArtistRecipe(
    id,
    prompt,
    artists,
    basePrompt: prompt,
    artistPrompt: prompt,
    pairId: id,
    variant: 'plain',
  );
}

/// Normalizes every valid prompt tag into explicit NovelAI numerical syntax.
V5ArtistWeightRepairResult normalizeV45ArtistSyntax(String input) =>
    _createRepairResult(_parsePromptTags(input));

/// Applies an independent community-derived multiplier in [1/3, 1/2] to every
/// valid tag. Only artist classification controls whether artist: is added.
V5ArtistWeightRepairResult repairV45ArtistWeightsForV5(
  String input, {
  double Function()? random,
}) {
  final secureRandom = random == null ? Random.secure() : null;
  final randomValue = random ?? secureRandom!.nextDouble;
  final tags = _migratePromptTags(_parsePromptTags(input), randomValue);
  return _createRepairResult(tags);
}

/// Creates several independently repaired complete strings for visual A/B
/// generation. Every candidate uses the same parser and migration rule.
List<ArtistRecipe> repairV45ArtistCandidatesForV5({
  required String input,
  required int count,
  int? drawSeed,
}) {
  final sourceTags = _parsePromptTags(input);
  if (sourceTags.isEmpty || count < 1) return const [];
  final random = Random(drawSeed);
  return List.generate(
    count,
    (index) => _recipeFromTags(
      _migratePromptTags(sourceTags, random.nextDouble),
      index,
      'v5-repair',
    ),
  );
}

({double min, double max}) _drawBounds(double first, double second) {
  final safeFirst = first.isFinite ? first.clamp(.05, 10).toDouble() : .2;
  final safeSecond = second.isFinite ? second.clamp(.05, 10).toDouble() : 1.2;
  return safeFirst <= safeSecond
      ? (min: safeFirst, max: safeSecond)
      : (min: safeSecond, max: safeFirst);
}

int countV5PromptTags(String input) => _parsePromptTags(input).length;

/// Rerolls every valid tag independently and retains the complete tag set in
/// every candidate. This is a weight draw, not a subset draw.
List<ArtistRecipe> drawAllV5ArtistWeights({
  required String input,
  required int count,
  double minWeight = defaultV5ArtistDrawMin,
  double maxWeight = defaultV5ArtistDrawMax,
  int? drawSeed,
}) {
  final sourceTags = _parsePromptTags(input);
  if (sourceTags.isEmpty || count < 1) return const [];

  final bounds = _drawBounds(minWeight, maxWeight);
  final random = Random(drawSeed);
  return List.generate(count, (index) {
    final tags = _migratePromptTags(
      sourceTags,
      random.nextDouble,
      bounds: bounds,
    );
    return _recipeFromTags(tags, index, 'v5-weight-${drawSeed ?? 0}');
  });
}

File: src/v5-artist-weight-repair.ts
import {
  canonicalArtistTagName,
  type GeneratedArtistRecipe,
  type ParsedRecipeToken,
  type RecipeTokenKind,
} from "./artist-recipe";
import { CURATED_ARTIST_TAGS } from "./curated-artists";

/**
 * Community migration reports for early NAI Diffusion V5 commonly suggest
 * starting at roughly one third to one half of a V4.5 prompt weight. NovelAI
 * does not publish an official conversion ratio, so this remains an
 * experimental migration heuristic rather than a model guarantee.
 */
export const MIN_V5_ARTIST_REPAIR_MULTIPLIER = 1 / 3;
export const MAX_V5_ARTIST_REPAIR_MULTIPLIER = 1 / 2;
export const DEFAULT_V5_ARTIST_DRAW_MIN = 0.2;
export const DEFAULT_V5_ARTIST_DRAW_MAX = 1.2;

export interface V5ArtistWeightRepairResult {
  output: string;
  /** Legacy field: number of parsed tags that already carried a weight. */
  adjustedWeightedGroups: number;
  /** Legacy field retained for compatibility with earlier callers. */
  wrappedUnweightedArtists: number;
  /** Every valid tag is normalized and adjusted, not just artist tags. */
  totalAdjusted: number;
  artistTagCount: number;
  qualityTagCount: number;
  otherTagCount: number;
}

interface PromptTag extends ParsedRecipeToken {
  sourceWeighted: boolean;
}

const QUALITY_PATTERN = /^(?:masterpiece|best[_ ]quality|amazing[_ ]quality|very[_ ]aesthetic|extremely[_ ]detailed(?:[_ ]cg)?|ultra[-_ ]?detailed|high[_ ]quality|great[_ ]quality|good[_ ]quality|average[_ ]quality|low[_ ]quality|worst[_ ]quality|aesthetic|very[_ ]pleasing|no[_ ]text)$/i;
const YEAR_PATTERN = /^year[_ ]?\d{4}$/i;
const NEGATIVE_PATTERN = /^(?:no\s+|negative\s+|avoid\s+)/i;
const STYLE_PATTERN = /(?:style|realism|impasto|illustration|painting|lineart|lighting|shading|\bcg\b|photorealistic|monochrome|sketch|watercolor|brushstroke|anime coloring|game cg)/i;
const NUMERIC_SCOPE_PATTERN = /([+-]?(?:\d+(?:\.\d*)?|\.\d+))\s*::([\s\S]*?)::/g;

// A bare token is inherently ambiguous. Only locally reviewed Danbooru
// category-1 names (plus the explicitly confirmed compatibility addition) are
// allowed to gain an artist: prefix automatically. Unknown bare tags still
// participate in repair/draw, but remain ordinary tags.
const KNOWN_BARE_ARTIST_TAGS = new Set([
  ...CURATED_ARTIST_TAGS.map((artist) => artist.name),
  "xiaoluo_xl",
]);

export function formatV5ArtistWeight(value: number) {
  const rounded = Math.round(value * 100) / 100;
  return (Object.is(rounded, -0) ? 0 : rounded)
    .toFixed(2)
    .replace(/\.00$/, "")
    .replace(/(\.\d)0$/, "$1");
}

function normalizePromptSource(input: string) {
  return input
    .normalize("NFKC")
    .replace(/\u00a0/g, " ")
    .replace(/\\([_()[\]{}])/g, "$1")
    .replace(/[，、；;\r\n]+/g, ",");
}

function cleanPromptValue(raw: string) {
  return raw
    .trim()
    .replace(/^(?:::)+\s*/, "")
    .replace(/\s*(?:::)+$/, "")
    .trim();
}

function repairKnownBareArtistName(raw: string): string | null {
  const candidate = canonicalArtistTagName(raw);
  if (KNOWN_BARE_ARTIST_TAGS.has(candidate)) return candidate;

  // Recover a single missing final parenthesis only when the repaired value
  // exactly matches the reviewed artist set. This fixes yd_(orange_maru
  // without guessing arbitrary malformed content tags.
  const opens = (candidate.match(/\(/g) ?? []).length;
  const closes = (candidate.match(/\)/g) ?? []).length;
  if (opens === closes + 1 && KNOWN_BARE_ARTIST_TAGS.has(`${candidate})`)) {
    return `${candidate})`;
  }
  return null;
}

function classifyAndNormalizeValue(raw: string): { value: string; kind: RecipeTokenKind } | null {
  const value = cleanPromptValue(raw);
  if (!value) return null;

  const explicitArtist = value.match(/^artist\s*:\s*([\s\S]+)$/i);
  if (explicitArtist) {
    const rawName = explicitArtist[1].trim();
    const name = repairKnownBareArtistName(rawName) ?? canonicalArtistTagName(rawName);
    return name ? { value: `artist:${name}`, kind: "artist" } : null;
  }

  const knownArtist = repairKnownBareArtistName(value);
  if (knownArtist) return { value: `artist:${knownArtist}`, kind: "artist" };

  const comparable = value.replace(/_/g, " ").trim();
  if (YEAR_PATTERN.test(value)) return { value, kind: "year" };
  if (QUALITY_PATTERN.test(value)) return { value, kind: "quality" };
  if (NEGATIVE_PATTERN.test(comparable)) return { value, kind: "negative" };
  if (STYLE_PATTERN.test(comparable)) return { value, kind: "style" };
  return { value, kind: "other" };
}

function parseTerm(
  raw: string,
  inheritedWeight: number | null,
  legacyLevel: number,
): PromptTag | null {
  let value = cleanPromptValue(raw);
  if (!value) return null;

  let baseWeight = inheritedWeight ?? 1;
  let sourceWeighted = inheritedWeight !== null || legacyLevel !== 0;

  // Stable-Diffusion-style compatibility form. Match the final numeric suffix
  // so artist names containing parentheses remain intact.
  const parenthesizedWeight = value.match(
    /^\(\s*([\s\S]+)\s*:\s*([+-]?(?:\d+(?:\.\d*)?|\.\d+))\s*\)$/,
  );
  if (parenthesizedWeight) {
    value = parenthesizedWeight[1].trim();
    baseWeight = Number(parenthesizedWeight[2]);
    sourceWeighted = true;
  } else if (value.startsWith("(") && value.endsWith(")")) {
    value = value.slice(1, -1).trim();
  }

  const normalized = classifyAndNormalizeValue(value);
  if (!normalized) return null;
  const weight = baseWeight * Math.pow(1.05, legacyLevel);
  return {
    raw,
    value: normalized.value,
    weight: Number.isFinite(weight) ? weight : 1,
    kind: normalized.kind,
    sourceWeighted,
  };
}

/**
 * Split a comma-delimited section while carrying NovelAI brace/square-bracket
 * emphasis across commas. Braces are removed because every output tag is
 * emitted using explicit numerical emphasis.
 */
function parseTermList(source: string, inheritedWeight: number | null): PromptTag[] {
  const output: PromptTag[] = [];
  let token = "";
  let curly = 0;
  let square = 0;
  let startingLevel: number | null = null;

  const flush = () => {
    const parsed = parseTerm(token, inheritedWeight, startingLevel ?? 0);
    if (parsed) output.push(parsed);
    token = "";
    startingLevel = null;
  };

  for (const char of source) {
    if (char === "{") {
      curly += 1;
      continue;
    }
    if (char === "}") {
      curly = Math.max(0, curly - 1);
      continue;
    }
    if (char === "[") {
      square += 1;
      continue;
    }
    if (char === "]") {
      square = Math.max(0, square - 1);
      continue;
    }
    if (char === ",") {
      flush();
      continue;
    }
    if (startingLevel === null && !/\s/.test(char)) {
      startingLevel = curly - square;
    }
    token += char;
  }
  flush();
  return output;
}

function parsePromptTags(input: string): PromptTag[] {
  const source = normalizePromptSource(input);
  const output: PromptTag[] = [];
  let cursor = 0;

  for (const match of source.matchAll(NUMERIC_SCOPE_PATTERN)) {
    const index = match.index ?? 0;
    output.push(...parseTermList(source.slice(cursor, index), null));
    const weight = Number(match[1]);
    output.push(...parseTermList(match[2], Number.isFinite(weight) ? weight : 1));
    cursor = index + match[0].length;
  }
  output.push(...parseTermList(source.slice(cursor), null));
  return output;
}

function formatPromptTags(tags: readonly Pick<PromptTag, "value" | "weight">[]) {
  return tags
    .map((tag) => `${formatV5ArtistWeight(tag.weight)}::${tag.value} ::`)
    .join(", ");
}

function createRepairResult(tags: PromptTag[]): V5ArtistWeightRepairResult {
  const artistTagCount = tags.filter((tag) => tag.kind === "artist").length;
  const qualityTagCount = tags.filter((tag) => tag.kind === "quality").length;
  return {
    output: formatPromptTags(tags),
    adjustedWeightedGroups: tags.filter((tag) => tag.sourceWeighted).length,
    wrappedUnweightedArtists: tags.filter(
      (tag) => tag.kind === "artist" && !tag.sourceWeighted,
    ).length,
    totalAdjusted: tags.length,
    artistTagCount,
    qualityTagCount,
    otherTagCount: tags.length - artistTagCount - qualityTagCount,
  };
}

function normalizedRandomSample(random: () => number) {
  const value = random();
  return Number.isFinite(value) ? Math.max(0, Math.min(1, value)) : 0.5;
}

function randomRepairMultiplier(random: () => number) {
  return MIN_V5_ARTIST_REPAIR_MULTIPLIER
    + normalizedRandomSample(random)
      * (MAX_V5_ARTIST_REPAIR_MULTIPLIER - MIN_V5_ARTIST_REPAIR_MULTIPLIER);
}

function migratePromptTags(
  tags: readonly PromptTag[],
  random: () => number,
  bounds?: readonly [number, number],
) {
  return tags.map((tag) => {
    const migrated = tag.weight * randomRepairMultiplier(random);
    const bounded = bounds
      ? Math.max(bounds[0], Math.min(bounds[1], migrated))
      : migrated;
    return {
      ...tag,
      weight: Math.round(bounded * 100) / 100,
    };
  });
}

function generatedRecipeFromTags(
  tags: readonly PromptTag[],
  index: number,
  idPrefix: string,
): GeneratedArtistRecipe {
  const prompt = formatPromptTags(tags);
  const artists = tags
    .filter((tag) => tag.kind === "artist")
    .map((tag) => ({
      name: canonicalArtistTagName(tag.value.replace(/^artist\s*:/i, "")),
      weight: tag.weight,
    }));
  const auxiliary: ParsedRecipeToken[] = tags
    .filter((tag) => tag.kind !== "artist")
    .map((tag) => ({
      raw: `${formatV5ArtistWeight(tag.weight)}::${tag.value} ::`,
      value: tag.value,
      weight: tag.weight,
      kind: tag.kind,
    }));
  const fingerprint = tags
    .map((tag) => `${tag.value}@${formatV5ArtistWeight(tag.weight)}`)
    .join("+");
  return {
    id: `${idPrefix}-${index + 1}-${Date.now().toString(36)}-${fingerprint}`,
    artists,
    auxiliary,
    mutations: [],
    franchiseStyles: [],
    basePrompt: prompt,
    prompt,
  };
}

/** Normalize every valid prompt tag to explicit NovelAI numerical syntax. */
export function normalizeV45ArtistSyntax(input: string): V5ArtistWeightRepairResult {
  return createRepairResult(parsePromptTags(input));
}

/**
 * Apply an independent [1/3, 1/2] multiplier to every valid tag. Artist tags
 * and quality/style/content tags all participate; only artist classification
 * controls whether the artist: prefix is added.
 */
export function repairV45ArtistWeightsForV5(
  input: string,
  random: () => number = Math.random,
): V5ArtistWeightRepairResult {
  const tags = migratePromptTags(parsePromptTags(input), random);
  return createRepairResult(tags);
}

/**
 * Create several independently repaired complete strings for visual A/B
 * testing. Every candidate uses the exact same parser and migration rule as
 * `repairV45ArtistWeightsForV5`; only the sampled multipliers differ.
 */
export function repairV45ArtistCandidatesForV5(
  input: string,
  count: number,
  random: () => number = Math.random,
): GeneratedArtistRecipe[] {
  const sourceTags = parsePromptTags(input);
  if (sourceTags.length === 0 || count < 1) return [];
  const total = Math.max(1, Math.floor(Number.isFinite(count) ? count : 1));
  return Array.from({ length: total }, (_, index) => generatedRecipeFromTags(
    migratePromptTags(sourceTags, random),
    index,
    "v5-repair",
  ));
}

function normalizedDrawBounds(rawMin: number, rawMax: number): [number, number] {
  const left = Number.isFinite(rawMin) ? rawMin : DEFAULT_V5_ARTIST_DRAW_MIN;
  const right = Number.isFinite(rawMax) ? rawMax : DEFAULT_V5_ARTIST_DRAW_MAX;
  const min = Math.max(0.05, Math.min(10, Math.min(left, right)));
  const max = Math.max(min, Math.max(0.05, Math.min(10, Math.max(left, right))));
  return [min, max];
}

/** Return the number of valid tags retained in every weight-draw candidate. */
export function countV5PromptTags(input: string) {
  return parsePromptTags(input).length;
}

/**
 * Migrate every valid tag with the same independent one-third-to-one-half rule
 * as the repair tool, then constrain the result to the user-selected draw
 * bounds. The original weight therefore remains meaningful (for example a
 * legacy weight of 2 becomes roughly 0.67-1.0), while the bounds provide a
 * final safety floor/ceiling. This is a complete-string weight draw, not a
 * subset draw.
 */
export function drawAllV5ArtistWeights(
  input: string,
  count: number,
  minWeight = DEFAULT_V5_ARTIST_DRAW_MIN,
  maxWeight = DEFAULT_V5_ARTIST_DRAW_MAX,
  random: () => number = Math.random,
): GeneratedArtistRecipe[] {
  const sourceTags = parsePromptTags(input);
  if (sourceTags.length === 0 || count < 1) return [];

  const [min, max] = normalizedDrawBounds(minWeight, maxWeight);
  const total = Math.max(1, Math.floor(Number.isFinite(count) ? count : 1));
  return Array.from({ length: total }, (_, index) => {
    const tags = migratePromptTags(sourceTags, random, [min, max]);
    return generatedRecipeFromTags(tags, index, "v5-weight-draw");
  });
}
